### PR TITLE
[WIP] revamp docs

### LINKED ISSUE #9: Section on Coscript object and its methods
in general sort of relating to #8, general underlying things that we should watch out for. Things like setShouldKeepAround and how it works, or MORuntime, etc

### LINKED ISSUE #15: Detail MSActionController & MSBaseAction(s)
These are two very useful APIs, as they allow you to invoke default functionality and enter modes otherwise inaccessible. I've guessed their usage based on header dumps and playing around, but official documentation would be nice :)

### LINKED ISSUE #32: MSImageProxy no longer exists
According to this class dump:
https://github.com/abynim/Sketch-Headers/tree/master/Headers

And this error message:
`2/9/16 11:59:35.229 AM Sketch[34925]: Exception: ReferenceError: Can't find variable: MSImageProxy
`

Although the docs believe it still walks among us:
http://developer.sketchapp.com/reference/MSImageProxy/

### LINKED ISSUE #72: Request:  Example of development workflow
Can you add an example of the workflow involved with creating and editing a sketch plugin.  Do I have to restart Sketch every time I edit the source of a plugin?  Or is there a way to force sketch to reload the plugin automatically?  Is there a command line for forcing sketch to reload plugins?  Something I could add to a GULP script's "watch" routine?

### LINKED ISSUE #109: missing documentation
Hi, I am writing a plugin but it seems some documentation is missing.

for example, where can i find some native api for sketch?

such as: http://developer.sketchapp.com/reference/class/MSTextLayer/ , it says 404.

Where can i find these native documents? Thanks a lot.

### PR DESCRIPTION
**Pull request by mathieudutour**
Fix #9, Fix #15, Fix #32, Fix #72, Fix #109

- [x] optimize sketch logo svg in navbar
- [x] rename `introduction` to `Guides`
- [x] Add ["First Plugin"](https://github.com/BohemianCoding/developer.sketchapp.com/pull/119/files#diff-db97deeba0c041166d684e99f5fe9ea4) tutorial
- [x] Add warning about [using the internal](https://github.com/BohemianCoding/developer.sketchapp.com/pull/119/files#diff-2ca97d02c804c2185410903b74e9bdb1) (fixes https://github.com/BohemianCoding/Sketch/issues/4496)
- [x] revamp the [first page](https://github.com/BohemianCoding/developer.sketchapp.com/pull/119/files#diff-3ff0627b4a78b86bb3d6ecf19a1580d1)
- [ ] add "next steps" everywhere
- [x] revamp CocoaScript [page](https://github.com/BohemianCoding/developer.sketchapp.com/pull/119/files#diff-fa3e48141d23b4994c6637fccf0cde43)
- [x] rewrite "Updating Plugin" to be "Publishing Plugin"
- [ ] revamp Dev setup page
- [x] add [page](https://github.com/BohemianCoding/developer.sketchapp.com/pull/119/files#diff-94ecf1eac7ab6dc5bbc21280e9b6df40) on "actions"
- [x] add debugging page

### CODE REVIEW

File: pages/index.md [hunk 1]
```diff
@@ -1,19 +1,17 @@
...
 ---
 
-Welcome to Sketch's Developer Site. We’ve worked very hard to make Sketch the dream “designer toolbox”, but everyone has slightly different needs, and there is a good chance that there’s a feature that you would like, which we’ve not implemented.
+We’ve worked very hard to make Sketch the dream “designer toolbox”, but everyone has slightly different needs, and there is a good chance that there’s a feature that you would like, which we’ve not implemented.
 
-Luckily, Sketch includes a plugin engine so that developers can extend its functionality.
+If you are interested in extending Sketch, you are in the right place. Here we present an outline of the Sketch extensibility documentation and how to quickly build your first Sketch plugin.
 
-For example, there are already plugins out there to do such diverse things as [design specs](https://github.com/utom/sketch-measure), [content generation](https://github.com/timuric/Content-generator-sketch-plugin) and [perspective transformations](https://github.com/jamztang/MagicMirror).
+If you just want to use existing plugin, see [Plugin Marketplace](https://sketchapp.com/extensions/plugins/).
```
Comment (bomberstudios) [inline pages/index.md]: We don't use the term 'Marketplace' to refer to the directory / catalog. Let's find something else : )

File: pages/index.md [hunk 2]
```diff
@@ -24,24 +22,36 @@ Plugins in Sketch can do anything a user can do. For example:
...
 - Automate existing features by calling menu options from plugins
+- [design specs](https://github.com/utom/sketch-measure)
+- [content generation](https://github.com/timuric/Content-generator-sketch-plugin)
+- [perspective transformations](https://github.com/jamztang/MagicMirror)
+
+The easiest way to see Sketch plugins in action is via the [Plugin Marketplace](https://sketchapp.com/extensions/plugins/). You can browse for useful plugins, install them to try them out and get an idea how you might extend Sketch for your own design scenarios.
+
+### Writing an Extension
+
+We created a small toolchain which makes it very easy to create a new plugin. It is great for starting out and you can also find existing plugin [examples](/examples/).
```
Comment (bomberstudios) [inline pages/index.md]: Did we? 🤔

File: pages/index.md [hunk 3]
```diff
@@ -24,24 +22,36 @@ Plugins in Sketch can do anything a user can do. For example:
...
+
+Extensions can be written in either TypeScript or JavaScript. VS Code offers a first class extension development experience where you can develop, build, run, test and debug all from within VS Code itself.
```
Comment (bomberstudios) [inline pages/index.md]: We should refer to the code editor by its full name (Visual Studio Code) and provide a link to it, _if we want to promote it as the official tool to use for plugin development_ (which, AFAIK is not something we have discussed? Or is this something that happened while I was out?)

File: pages/index.md [hunk 4]
```diff
@@ -24,24 +22,36 @@ Plugins in Sketch can do anything a user can do. For example:
...
-If you want to learn how to do that, you’ve come to the right place.
+-->
 
-<div class="embed-container">
-  <iframe src="https://www.youtube.com/embed/TlSfoGN6WRw?rel=0&amp;showinfo=0&amp;color=white" frameborder="0" allowfullscreen=""></iframe>
-</div>
+### Extension ideas
 
-## Sections
+Lots of great community ideas for Sketch features are better implemented as plugins rather than as part of the core product. This way users can easily pick and choose the functionality they want, by installing the right set of plugins. The Sketch team tracks possible plugin as GitHub issues on the [plugin-request repository](https://github.com/sketchplugins/plugin-requests/issues). If you're looking for a great plugin to build, have a look at the issues.
```
Comment (bomberstudios) [inline pages/index.md]: Not sure what we're trying to say here, but we're definitely _not_ keeping track of requests on that repo. What is the message we want to send?

File: pages/index.md [hunk 5]
```diff
@@ -24,24 +22,36 @@ Plugins in Sketch can do anything a user can do. For example:
...
+* [Extension API](/reference/) - Learn about the Sketch extensibility APIs.
+* [Extension Examples](/examples/) - A list of extension samples you can review and build.
 
-## Help us improve!
+### Help us improve!
 
-If you find any error or omission on the documentation, or you’d like us to cover or clarify something, just [file an issue](https://github.com/BohemianCoding/developer.sketchapp.com/issues) and we’ll try to fix it. Of course, since all the content on this site is open source, you can help us improve by [suggesting an edit on GitHub]({{site.github_repo}}) (there's also an "Improve this page" link at the bottom of every page, in case you find something wrong while browsing the site).
+If you find any error or omission on the documentation, or you’d like us to cover or clarify something, just [file an issue]({{site.github_repo}}/issues) and we’ll try to fix it. Of course, since all the content on this site is open source, you can help us improve by [suggesting an edit on GitHub]({{site.github_repo}}) (there's also an "Improve this page" link at the bottom of every page, in case you find something wrong while browsing the site).
```
Comment (bomberstudios) [inline pages/index.md]: `{{site.github_repo}}` points to `http://192.168.1.36:3000/` here, because it's not defined in `_config.yml`. Is it defined as an environment variable when we deploy to GitHub Pages? Otherwise we'll need to define it in the config.

File: pages/reference.md [hunk 6]
```diff
@@ -5,7 +5,23 @@ permalink: /reference/
 script: /js/search.js
 ---
 
-The pages below contain brief descriptions of all the actions that your plugin can listen for, and some of the key Sketch classes that they can interact with.
+Sketch gives you full access to its internals. This gives you immense power (and with great power comes great responsibility!).
+
+That doesn't mean that we are gonna freeze our codebase and we do change the intervals regularly. As a result you need to accept that your plugins may well break regularly.
```
Comment (bomberstudios) [inline pages/reference.md]: intervals → internals

I understand what we're trying to say here, but I'd like to reword it to sound a bit less catastrophic : )

What about:

> The plugin system in Sketch gives you full access to the app's internals and the core frameworks in macOS. So you have an immense power to build almost _anything_.
>
> With great power comes great responsibility, tough, so you'll need to keep an eye on your code with each Sketch release. We change Sketch's internals from time to time when refactoring, and as a result your plugins may be calling some methods that were renamed or removed.
>
> We'll do our best to warn you about that, but it doesn't hurt to download [our beta](http://sketchapp.com/beta/) to make sure your plugins will still work with new releases.

File: pages/reference.md [hunk 7]
```diff
@@ -5,7 +5,23 @@ permalink: /reference/
...
 - [Javascript API](/reference/api)
 - [Actions](/reference/action)
+
+Even though we are not going to document the internals, there are 3 sources of information you can look into:
+
+- [The official AppKit document](https://developer.apple.com/documentation/appkit?language=objc): this is Apple framework Sketch is built upon.
+- [Foundation](https://developer.apple.com/documentation/foundation?language=objc): more essential Apple classes and services.
+- [The Sketch Headers](https://github.com/abynim/Sketch-Headers) (Thanks @abynim 🙏): this is the headers of all the classes used by Sketch. If your plugin breaks with a new release because you used a method that was removed, you can check the diffs to find a replacement.
+
+Again, the last link is to use at your own risk, we are not going to document are freeze any of this but we want to give you the power to do anything.
```
Comment (bomberstudios) [inline pages/reference.md]: document are freeze → document *or* freeze ?

File: blog/index.html [hunk 8]
```diff
@@ -3,6 +3,9 @@
 layout: blog-index
 ---
 <section class="blog">
+  <div style="margin-bottom: 50px;">
+    <em>You will find the old developer blog here. The new articles are now there: <a href="https://blog.sketchapp.com" target="_blank">https://blog.sketchapp.com</a>.</em>
```
Comment (bomberstudios) [inline blog/index.html]: Have you confirmed this is the case? We're not posting anything developer-related in Sketch's blog, and even if the blog here is definitely neglected, we haven't made the decision to kill it, AFAIK (but I could be wrong, since I've definitely missed some discussions lately!)

File: _sass/components/_navigation-toc.scss [hunk 9]
```diff
@@ -50,6 +50,13 @@
   padding: 0
 }
 
+.toc-nav .separator {
```
Comment (bomberstudios) [inline _sass/components/_navigation-toc.scss]: ping @intemperie about this change. Last time we spoke, we were trying to keep some sort of consistency between the public and the developer site, so maybe this needs to be ported across.

File: _sass/base/_typography.scss [hunk 10]
```diff
@@ -371,7 +371,7 @@ pre, code {
 }
 
 code {
-  padding: .0625em .25em; //1px 4px
+  padding: .25em .375em; //1px 4px
```
Comment (bomberstudios) [inline _sass/base/_typography.scss]: ping @intemperie about this change. Last time we spoke, we were trying to keep some sort of consistency between the public and the developer site, so maybe this needs to be ported across.

File: _layouts/sidebar-page.html [hunk 11]
```diff
@@ -11,10 +11,14 @@ <h1 class="h3 nav-toggle">Contents</h1>
...
           <ol>
-            {% if page.url contains '/introduction/' %}
-                {% assign introductions = site.introduction | sort:"order" %}
-                {% for introduction in introductions %}
-                  <li><a href="{{introduction.url}}" {% if page.url == introduction.url %} class="active"{% endif %}>{{introduction.title}}</a></li>
+            {% if page.url contains '/guides/' %}
+              <li class="separator">Getting Started</li>
+                {% assign guides = site.guides | sort:"order" %}
+                {% for guide in guides %}
+                  {% if guide.order == 200 %}
```
Comment (bomberstudios) [inline _layouts/sidebar-page.html]: I'm sure there's a better way to do this. I can live with it by now, but please promise we'll improve it in the future : )

File: _guides/publishing.md [hunk 12]
```diff
@@ -1,10 +1,12 @@
...
-title: Updating Plugins
-summary: How to setup a plugin to be updated by Sketch
-permalink: /introduction/updating-plugins/
+title: Publishing Plugins
+summary: How to publish a plugin for the first and how to publish updates
+permalink: /guides/publishing-plugins/
 order: 150
 ---
 
+// TODO talk about the plugin directory and how to add one here. Talk about `skpm publish`.
```
Comment (bomberstudios) [inline _guides/publishing.md]: This is not a comment, and it's showing up in the published page 😬

File: _guides/preferences.md [hunk 13]
```diff
@@ -1,8 +1,8 @@
 ---
-title: Scripting Preferences
+title: Dev Environment
```
Comment (bomberstudios) [inline _guides/preferences.md]: I'm not entirely in love with the title change, but I understand where this is coming from.

Do we have alternatives for this title?

File: _guides/actions.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
++    "identifier" : "my-action-listener-identifier"
++  }
+  ...
+],
+```
+
+We're telling our plugin that we want to run the `onOpenDocument` function when a document is open, so let's add that in `my-action-listener.js`:
+
+```js
+export function onOpenDocument (context) {
```
Comment (bomberstudios) [inline _guides/actions.md]: This assumes we're using `skpm`, maybe we can simplify this example, for the sake of clarity?

File: _guides/first-plugin.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+## The plugin manifest: manifest.json
+
+* Please read the manifest.json [manifest reference](/guides/plugin-bundles/#manifest)
+* Each Sketch plugin must have a manifest.json file that describes it and its capabilities.
+* Sketch reads this file during start-up.
+
+## Debugging your Extension
+
+// TODO
```
Comment (bomberstudios) [inline _guides/first-plugin.md]: No `// TODO` stuff in production, please : )

File: _guides/actions.md (hunk 14 above)
Comment (mathieudutour) [reply _guides/actions.md]: That's exactly what we need to decide. Either we decide to go full-on `skpm` or not but if we don't, nothing will be clear because `var onRun = function (context) {}` is really unclear.

I think that people who do not want to use `skpm` can understand it by themselves. We just want to make sure that people getting into it can understand, which means picking the easiest way.

There should be another document, probably living in the skpm documentation, explaining what it does behind the scene so that people can do it by themselves if they choose to.

File: _guides/first-plugin.md (hunk 15 above)
Comment (mathieudutour) [reply _guides/first-plugin.md]: yeah there are a few tasks to finish before merging (see PR description), I just wanted to get your early thoughts on this

File: _guides/preferences.md (hunk 13 above)
Comment (mathieudutour) [reply _guides/preferences.md]: That's because I want to expand this section with some `skpm` explanation (hence, not only "Scripting Preferences" but more "Development Environment"

Comment (mathieudutour) [reply _guides/preferences.md]: or "Dev Setup", or "Dev tools"

File: _guides/publishing.md (hunk 12 above)
Comment (mathieudutour) [reply _guides/publishing.md]: see comment about WIP ;)

File: _layouts/sidebar-page.html (hunk 11 above)
Comment (mathieudutour) [reply _layouts/sidebar-page.html]: yeah I'm not super happy with it either

File: blog/index.html (hunk 8 above)
Comment (mathieudutour) [reply blog/index.html]: oh I wasn't sure. It feels weird to have 2 different platforms for it tho 🤔

File: _sass/base/_typography.scss (hunk 10 above)
Comment (intemperie) [reply _sass/base/_typography.scss]: Thanks for the ping. Please, don't touch the styles based on personal preferences :P

BTW, we talked about moving the Dev site to the main repo at some point, so CSS styles are global/shared for everything. What happened to that idea?

File: pages/index.md (hunk 2 above)
Comment (mathieudutour) [reply pages/index.md]: yes, `skpm` ;)

File: _sass/components/_navigation-toc.scss (hunk 9 above)
Comment (intemperie) [reply _sass/components/_navigation-toc.scss]: What's this exactly? I'll need to build the branch myself and see, but I'm quite busy ATM.

File: _sass/base/_typography.scss (hunk 10 above)
Comment (mathieudutour) [reply _sass/base/_typography.scss]: can we change this then? It's pretty unreadable without more spacing

File: pages/index.md (hunk 4 above)
Comment (mathieudutour) [reply pages/index.md]: That's what is writing on the repo tho. I wanted to make sure people don't come in this repo with plugin requests. See https://github.com/BohemianCoding/developer.sketchapp.com/issues/112

File: pages/index.md (hunk 5 above)
Comment (mathieudutour) [reply pages/index.md]: it's used just below ;)

File: pages/reference.md (hunk 6 above)
Comment (mathieudutour) [reply pages/reference.md]: hum not sure about "We'll do our best to warn you about that", it kind of feel like an engagement that we probably don't want to make?

File: _sass/components/_navigation-toc.scss (hunk 9 above)
Comment (mathieudutour) [reply _sass/components/_navigation-toc.scss]: It is the list separator ("GETTING STARTED" and "ADVANCED"): 
screen shot 2017-10-23 at 17 12 17

File: _sass/base/_typography.scss (hunk 10 above)
Comment (intemperie) [reply _sass/base/_typography.scss]: That's a quite subjective affirmation. I don't see a readability issue. While the extra padding is nice, there is a reason for everything. `<kbd>` tags styling is too similar, and multiline `<code>` tags look cramped. Please leave it as it was.

Comment (mathieudutour) [reply _sass/base/_typography.scss]: haha yes, of course, it's subjective. There is no markdown markup for `<kbd>` so I'm stuck with `<code>` which looks really packed.

I'll roll it back but I really hope we can reconsider and make this website code friendly since it's entirely about it.

(
for reference:
with small padding:
screen shot 2017-10-23 at 17 26 33

with bigger padding:
screen shot 2017-10-23 at 17 22 43

I feel that the second one is a lot easier to parse visually but of course that's subjective :) )
)

File: _sass/components/_navigation-toc.scss (hunk 9 above)
Comment (intemperie) [reply _sass/components/_navigation-toc.scss]: There are styles for that already. Check https://sketchapp.com/docs/

File: _sass/base/_typography.scss (hunk 10 above)
Comment (intemperie) [reply _sass/base/_typography.scss]: Sure, there's always room for improvement! The problem is that we need to account for all the things first :)

A bit of context: We came from dark `<code>` tags, but the contrast was too high and many people didn't like it, so we ended up with these. You can see how inline `kbd` tags look now here: https://sketchapp.com/docs/the-interface/inspector/ (check Math section). 

Maybe we could improve the visual differentiation between the two, so we can have that extra padding, which is indeed nicer, but we still have the multiline "problem". Let me make a couple of tests, ok? :)

Comment (mathieudutour) [reply _sass/base/_typography.scss]: sure, it's really not a priority tho, I just did that because I thought it was a low hanging fruit :)

File: _guides/preferences.md (hunk 13 above)
Comment (bomberstudios) [reply _guides/preferences.md]: "Development Environment" sounds good : )

Comment (mathieudutour) [reply _guides/preferences.md]: Yeah but it doesn’t fit in the side bar haha

File: _guides/actions.md (hunk 14 above)
Comment (bomberstudios) [reply _guides/actions.md]: Ok, let's go with this then : )

**Comment by intemperie** [comment]
I slightly tweaked the `code` styles, so we don't have those issues you mentioned before. I hope you like it :)
